src/cargo/util/toml.rs

```diff
@@ -426,28 +427,44 @@ impl TomlManifest {
         let blacklist = vec!["build", "deps", "examples", "native"];
 
         for bin in bins.iter() {
-            if blacklist.iter().find(|&x| *x == bin.name) != None {
+            if blacklist.iter().find(|&x| *x == bin.name()) != None {
                 return Err(human(&format!("the binary target name `{}` is \
-                                           forbidden", bin.name)));
+                                           forbidden", bin.name())));
             }
         }
 
         let examples = match self.example {
-            Some(ref examples) => examples.clone(),
-            None => inferred_example_targets(layout),
+            Some(ref examples) => {
+                for target in examples {
+                    try!(validate_example_name(target));
+                }
+                examples.clone()
+            }
+            None => inferred_example_targets(layout)
         };
 
         let tests = match self.test {
-            Some(ref tests) => tests.clone(),
-            None => inferred_test_targets(layout),
+            Some(ref tests) => {
+                for target in tests {
+                    try!(validate_test_name(target));
+                }
+                tests.clone()
+            }
+            None => inferred_test_targets(layout)
         };
 
-        let benches = if self.bench.is_none() || self.bench.as_ref().unwrap().is_empty() {
-            inferred_bench_targets(layout)
-        } else {
-            self.bench.as_ref().unwrap().iter().map(|t| {
-                t.clone()
-            }).collect()
+        let benches = match self.bench {
+            Some(ref benches) => {
+                for target in benches {
+                    try!(validate_bench_name(target));
+                }
+                if benches.is_empty() {
```

Comment:
Could this actually pick up the same logic as the tests above? E.g. if this was explicitly specified as empty we shouldn't then to go try to infer anything.